File: src/controllers/warehouse.rs
use axum::{
    extract::{Path, State},
    response::IntoResponse,
    Json,
};
use uuid::Uuid;

use crate::{services, utils::db::DB, validations::warehouse::StoreWarehouseSchema};

pub async fn index(State(db): State<DB>) -> impl IntoResponse {
    let warehouses = services::warehouse::all(&db).await;
    Json(warehouses)
}

pub async fn show(Path(id): Path<Uuid>, State(db): State<DB>) -> impl IntoResponse {
    let warehouse = services::warehouse::find(&id, &db).await;
    Json(warehouse)
}

pub async fn store(
    State(db): State<DB>,
    Json(input): Json<StoreWarehouseSchema>,
) -> impl IntoResponse {
    services::warehouse::insert(&input, &db).await;
}

pub async fn update(
    Path(id): Path<Uuid>,
    State(db): State<DB>,
    Json(input): Json<StoreWarehouseSchema>,
) -> impl IntoResponse {
    services::warehouse::update(&id, &input, &db).await;
}

pub async fn destroy(Path(id): Path<Uuid>, State(db): State<DB>) -> impl IntoResponse {
    services::warehouse::delete(&id, &db).await;
}

File: apps/dashboard/app/services/warehouse.ts
import { env } from "@/env";
import { api } from "./auth";
import {
  warehouseSchema,
  StoreWarehousePayload,
} from "@/validations/warehouse";
import { z } from "zod";

class Warehouse {
  public async all(cookie?: string) {
    const url = new URL(env.VITE_API_URL);
    url.pathname = "/warehouses";

    const response = await api(url, {
      headers: cookie ? { cookie } : undefined,
    });

    const data = await response.json();

    return z.array(warehouseSchema).parse(data);
  }

  public async show(id: string, cookie?: string) {
    const url = new URL(env.VITE_API_URL);
    url.pathname = `/warehouses/${id}`;

    const response = await api(url, {
      headers: cookie ? { cookie } : undefined,
    });

    const data = await response.json();

    return warehouseSchema.parse(data);
  }

  public store(payload: StoreWarehousePayload, cookie?: string) {
    const url = new URL(env.VITE_API_URL);
    url.pathname = "/warehouses";

    return api(url, {
      method: "POST",
      headers: cookie ? { cookie } : undefined,
      body: JSON.stringify(payload),
    });
  }

  public update(id: string, payload: StoreWarehousePayload, cookie?: string) {
    const url = new URL(env.VITE_API_URL);
    url.pathname = `/warehouses/${id}`;

    return api(url, {
      method: "PATCH",
      headers: cookie ? { cookie } : undefined,
      body: JSON.stringify(payload),
    });
  }

  public destroy(id: string, cookie?: string) {
    const url = new URL(env.VITE_API_URL);
    url.pathname = `/warehouses/${id}`;

    return api(url, {
      method: "DELETE",
      headers: cookie ? { cookie } : undefined,
    });
  }
}

export const warehouseService = new Warehouse();
